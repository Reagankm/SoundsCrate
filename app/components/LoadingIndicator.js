/* @flow */
import React, { Component } from 'react';
import {
  Animated,
  Image,
  StyleSheet,
  View,
} from 'react-native';
import record from './assets/record.png';
import needle from './assets/needle.png';
import player from './assets/player.png';

const ANIMATION_DURATION = 500;

export default class LoadingIndicator extends Component {
  needleAnimation: Animated.Value = new Animated.Value(0);

  componentDidMount() {
    this.animateNeedle();
  }

  render(): React.Element<*> {
    const rotate = this._getInterpolation('0deg', '-30deg');
    const recordRotate = this._getInterpolation('0deg', '360deg');
    const translateX = this._getInterpolation(0, 6);
    const translateY = this._getInterpolation(0, 4);
    const transform = [{ rotate }, { translateX }, { translateY }];
    return (
      <View style={styles.root}>
        <Image style={styles.player} source={player} />
        <Animated.Image
          style={[styles.record, {
            transform: [{ rotate: recordRotate }] }
          ]} source={record} />
        <Animated.Image style={[styles.needle, { transform }]} source={needle} />
      </View>
    );
  }

  animateNeedle() {
    this._getNeedleAnimation(30).start(({ finished }: any) => {
      finished && this._getNeedleAnimation(0).start(({ finished }: any) => {
        finished && this.animateNeedle();
      });
    });
  }

  _getInterpolation(...outputRange: Array<string | number>): Animated.Value {
    return this.needleAnimation.interpolate({
      inputRange: [0, 30],
      outputRange,
    });
  }

  _getNeedleAnimation(toValue: number): Animated.Value {
    return Animated.spring(
      this.needleAnimation,
      { toValue,
        duration: ANIMATION_DURATION,
        useNativeDriver: true, }
    );
  }
}

const styles = StyleSheet.create({
  root: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  player: {
    position: 'absolute',
    width: 100,
    height: 100,
  },
  record: {
    position: 'absolute',
    width: 80,
    height: 80,
  },
  needle: {
    width: 20,
    height: 45,
    marginTop: -30,
    marginLeft: 65,
  },
});
