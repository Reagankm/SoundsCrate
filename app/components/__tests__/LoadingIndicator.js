/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import LoadingIndicator from '../LoadingIndicator';

jest.mock('Animated', () => {
  class Value {
    interpolate(config) { return config; }
  }
  return {
    Value,
    Image: 'Image',
    spring(config) { return { start: () => {} } },
  };
});
jest.mock('Image', () => 'Image');
jest.mock('View', () => 'View');

describe('app/components/LoadingIndicator', () => {
  it('should render', () => {
    expect(renderer.create(<LoadingIndicator />)).toMatchSnapshot();
  });
});
