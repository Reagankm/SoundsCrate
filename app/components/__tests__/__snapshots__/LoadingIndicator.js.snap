exports[`app/components/LoadingIndicator should render 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      undefined,
    ]
  }>
  <Image
    source={1}
    style={
      Object {
        "height": 100,
        "position": "absolute",
        "width": 100,
      }
    } />
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 80,
          "position": "absolute",
          "width": 80,
        },
        Object {
          "transform": Array [
            Object {
              "rotate": Object {
                "inputRange": Array [
                  0,
                  30,
                ],
                "outputRange": Array [
                  "0deg",
                  "360deg",
                ],
              },
            },
          ],
        },
      ]
    } />
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 45,
          "marginLeft": 65,
          "marginTop": -30,
          "width": 20,
        },
        Object {
          "transform": Array [
            Object {
              "rotate": Object {
                "inputRange": Array [
                  0,
                  30,
                ],
                "outputRange": Array [
                  "0deg",
                  "-30deg",
                ],
              },
            },
            Object {
              "translateX": Object {
                "inputRange": Array [
                  0,
                  30,
                ],
                "outputRange": Array [
                  0,
                  6,
                ],
              },
            },
            Object {
              "translateY": Object {
                "inputRange": Array [
                  0,
                  30,
                ],
                "outputRange": Array [
                  0,
                  4,
                ],
              },
            },
          ],
        },
      ]
    } />
</View>
`;
